packages/vite/src/node/plugins/define.ts

```diff
@@ -49,16 +57,20 @@ export function definePlugin(config: ResolvedConfig): Plugin {
   const importMetaKeys: Record<string, string> = {}
   const importMetaFallbackKeys: Record<string, string> = {}
   if (isBuild) {
-    env.SSR = !!config.build.ssr
-
+    const env: Record<string, any> = {
+      ...config.env,
+      SSR: !!config.build.ssr,
+    }
     // set here to allow override with config.define
     importMetaKeys['import.meta.hot'] = `undefined`
     for (const key in env) {
       importMetaKeys[`import.meta.env.${key}`] = JSON.stringify(env[key])
     }
```

Comment:
I split off `env` and `userDefineEnv` as otherwise the define envs get mixed into this for loop, which is harmless in practice, but would be nice to avoid.